pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/SegmentMetadataImpl.java

```diff
@@ -94,18 +96,46 @@
   private long _segmentStartTime;
   private long _segmentEndTime;
 
-  public SegmentMetadataImpl(File indexDir)
+  public SegmentMetadataImpl(@Nonnull File indexDir, @Nonnull SegmentVersion segmentVersion)
       throws ConfigurationException, IOException {
-    LOGGER.debug("SegmentMetadata location: {}", indexDir);
-    File tmpMetadataFile = SegmentDirectoryPaths.findMetadataFile(indexDir);
-    if (tmpMetadataFile == null) {
-      LOGGER.warn("Metadata file: {} does not exist in directory: {}", tmpMetadataFile, indexDir);
-      tmpMetadataFile = new File(indexDir, MetadataKeys.METADATA_FILE_NAME);
+    _metadataFile = SegmentDirectoryPaths.findMetadataFile(indexDir, segmentVersion);
+    Preconditions.checkNotNull(_metadataFile, "Cannot find segment metadata file under directory: " + indexDir);
+
+    _segmentMetadataPropertiesConfiguration = new PropertiesConfiguration(_metadataFile);
+    _columnMetadataMap = new HashMap<>();
+    _allColumns = new HashSet<>();
+    _schema = new Schema();
+
+    // For segment reload, we will load segment from the temporary index directory, but we need to set indexDir to the
+    // original index directory
+    String indexDirToLoad = indexDir.getPath();
+    if (indexDirToLoad.endsWith(RELOAD_TEMP_DIR_SUFFIX)) {
+      _indexDir = indexDirToLoad.substring(0, indexDirToLoad.length() - RELOAD_TEMP_DIR_SUFFIX.length());
+    } else {
+      _indexDir = indexDirToLoad;
+    }
+
+    init();
+    File creationMetaFile = SegmentDirectoryPaths.findCreationMetaFile(indexDir, segmentVersion);
+    if (creationMetaFile != null) {
+      loadCreationMeta(creationMetaFile);
     }
-    _metadataFile = tmpMetadataFile;
+
+    setTimeInfo();
+    _totalDocs = _segmentMetadataPropertiesConfiguration.getInt(V1Constants.MetadataKeys.Segment.SEGMENT_TOTAL_DOCS);
+    _totalRawDocs =
+        _segmentMetadataPropertiesConfiguration.getInt(V1Constants.MetadataKeys.Segment.SEGMENT_TOTAL_RAW_DOCS,
+            _totalDocs);
+  }
+
+  public SegmentMetadataImpl(File indexDir)
+      throws ConfigurationException, IOException {
```

Comment:
share code with above constructor by having common internal method taking metadataFile as param. (if java constructor calling rules allow that).

I would recommend static named methods to construct metadata but that would be bigger change touching lots of files.